reference/src/layout/unions.md

```diff
@@ -38,3 +28,79 @@ contents are.
 Even if the offsets happen to be all 0, there might still be differences in the
 function call ABI.  If you need to pass unions by-value across an FFI boundary,
 you have to use `#[repr(C)]`.
+
+### Layout of "repr C" unions
+
+The layout of `repr(C)` unions follows the C layout scheme. Per sections
+[6.5.8.5] and [6.7.2.1.16] of the C11 specification, this means that the offset
+of every field is 0. Unsafe code can cast a pointer to the union to a field type
+to obtain a pointer to any field, and vice versa. 
+
+[6.5.8.5]: http://port70.net/~nsz/c/c11/n1570.html#6.5.8p5
+[6.7.2.1.16]: http://port70.net/~nsz/c/c11/n1570.html#6.7.2.1p16
+
+#### Padding
+
+Since all fields are at offset 0, `repr(C)` unions do not have padding before
+their fields. They can, however, have trailing padding, to make sure the size is
+a multiple of the alignment:
+
+```rust
+# use std::mem::{size_of, align_of};
+#[repr(C, align(2))]
+union U { x: u8 }
+# fn main() {
+// The repr(align) attribute raises the alignment requirement of U to 2
+assert_eq!(align_of::<U>(), 2);
+// This introduces trailing padding, raising the union size to 2
+assert_eq!(size_of::<U>(), 2);
+# }
+```
+
+> **Note**: Fields are overlapped instead of laid out sequentially, so 
+> unlike structs there is no "between the fields" that could be filled 
+> with padding.
+
+#### Zero-sized fields
+
+If a `#[repr(C)]` union contains a field of zero-size, that field does not
+occupy space in the union. For example:
```

Comment:
Is it just me, or does this sentence effectively say "a field of size zero has size 0", and then even goes on with an example to demonstrate? That seems... silly. What am I missing?

You seem dead set on having some example about zero-sized fields in this section. Okay, fine. But why *two*, with one being... utterly trivial?

Reply:
> What am I missing?

Do we specify anywhere in the document what the size of the field or the union containing the field is?

We consider these types FFI-safe, so from there it can be derived that this has to have some well defined rules. AFAICT we only say "this behaves like C" but C does not have any zero-sized types, so that does not cover this use case. 

C implementations often give these types size 0, such that the union has size 0, but C++ for example gives these fields size 1, such that the union ends up having size 1. 

What does Rust do here for unions and where is that written down ?

Reply:
There's nothing special about size 0 though in the way to do layout. With something like https://github.com/rust-lang/unsafe-code-guidelines/issues/185, this falls out of a general-principle. We should explain the general principle of `repr(C)` union layout, not list each case (size 0, size 1, size 5, ...) separately.

You already added a remark covering the fact that this differs from what C++ does.

Reply:
> We should explain the general principle of repr(C) union layout, not list each case (size 0, size 1, size 5, ...) separately.

Sure, and part of this general principle is that `repr(C)` unions do not error when a zero-sized field is used, but that instead it actually treats them as a zero-sized field.

Reply:
> Sure, and part of this general principle is that repr(C) unions do not error when a zero-sized field is used, but that instead it actually treats them as a zero-sized field.

Right, so we should describe the general thing and then maybe just say "(this includes fields of size 0)", and that should be all there is specific to that size in the repr(Rust) and repr(C) sections. And then at the bottom of the page, something like: "what we do matches what GCC and clang do for unions in C, but note that in C++, structs with no fields size 1 but also behave a bit like zero-sized types in Rust; hence there are compatibility issues".

This PR heavily implies like zero-sized fields are somehow the exception from some rule *when considering Rust and C FFI*. They are not an exception though, they work like any other size. That makes it very confusing to treat them as an exception. They are an exception only in C++, there is no reason that exception has to taint anything but the compatibility remarks specific to that language.

Reply:
Check the latest changes out.